test/integration/apiserver/apiserver_test.go

```diff
@@ -389,59 +390,66 @@ func TestListOptions(t *testing.T) {
 	for _, watchCacheEnabled := range []bool{true, false} {
 		t.Run(fmt.Sprintf("watchCacheEnabled=%t", watchCacheEnabled), func(t *testing.T) {
 			tCtx := ktesting.Init(t)
+			prefix := path.Join("/", guuid.New().String(), "registry")
+			etcdConfig := storagebackend.Config{
+				Prefix:    prefix,
+				Transport: storagebackend.TransportConfig{ServerList: []string{framework.GetEtcdURL()}},
+			}
+			rawClient, kvClient, err := integration.GetEtcdClients(etcdConfig.Transport)
+			if err != nil {
+				t.Fatal(err)
+			}
+			// kvClient is a wrapper around rawClient and to avoid leaking goroutines we need to
+			// close the client (which we can do by closing rawClient).
+			defer func() {
+				err := rawClient.Close()
+				if err != nil {
+					t.Fatal(err)
+				}
+			}()
 
-			var storageTransport *storagebackend.TransportConfig
-			clientSet, _, tearDownFn := framework.StartTestServer(tCtx, t, framework.TestServerSetup{
-				ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
-					opts.Etcd.EnableWatchCache = watchCacheEnabled
-					storageTransport = &opts.Etcd.StorageConfig.Transport
-				},
-			})
-			defer tearDownFn()
-
-			ns := framework.CreateNamespaceOrDie(clientSet, "list-options", t)
-			defer framework.DeleteNamespaceOrDie(clientSet, ns, t)
-
-			rsClient := clientSet.AppsV1().ReplicaSets(ns.Name)
-
-			var compactedRv, oldestUncompactedRv string
+			var compactedRv string
+			var oldestUncompactedRv int64
 			for i := 0; i < 15; i++ {
-				rs := newRS(ns.Name)
+				rs := newRS("default")
 				rs.Name = fmt.Sprintf("test-%d", i)
-				created, err := rsClient.Create(tCtx, rs, metav1.CreateOptions{})
+				serializer := protobuf.NewSerializer(nil, nil)
+				buf := bytes.Buffer{}
+				err := serializer.Encode(rs, &buf)
+				if err != nil {
+					t.Fatal(err)
+				}
+				key := prefix + "/replicasets/default/" + rs.Name
+
+				resp, err := kvClient.Put(tCtx, key, buf.String())
 				if err != nil {
 					t.Fatal(err)
 				}
 				if i == 0 {
-					compactedRv = created.ResourceVersion // We compact this first resource version below
+					compactedRv = strconv.FormatInt(resp.Header.Revision, 10) // We compact this first resource version below
 				}
 				// delete the first 5, and then compact them
 				if i < 5 {
-					var zero int64
-					if err := rsClient.Delete(tCtx, rs.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero}); err != nil {
+					if _, err := kvClient.Delete(tCtx, key); err != nil {
 						t.Fatal(err)
 					}
-					oldestUncompactedRv = created.ResourceVersion
+					oldestUncompactedRv = resp.Header.Revision
 				}
 			}
-
-			// compact some of the revision history in etcd so we can test "too old" resource versions
-			rawClient, kvClient, err := integration.GetEtcdClients(*storageTransport)
+			_, err = kvClient.Compact(tCtx, int64(oldestUncompactedRv))
```

Comment:
This change makes sense to me, but it technically weakens this test for watchcache (i.e. compaction doesn't have to clear the cache, which technically was always the case).
So I'm wondering if we ever introduced any guarantees regarding compaction somewhere? I don't think, but just double checking.

@jpbetz - if you have thoughts

Reply:
> compaction doesn't have to clear the cache, which technically was always the case

Not following, watch cache is not aware of compaction.

Reply:
> So I'm wondering if we ever introduced any guarantees regarding compaction somewhere? I don't think, but just double checking.
> 
> @jpbetz - if you have thoughts

This change looks OK to me.  Since compaction can be initiated either directly by etcd or by client request,  I'm don't believe we can expect any guarantee that would cause a problem here.  It's always possible that the cache may serve data that is already compacted in etcd.